### DISCUSSION TITLE
Docker Swarm and Nginx + Varnish

### DISCUSSION THREAD

**Discussion by ghostlexly** [BretFisher/ama#119] [Q&A] [answered]
I have a docker-compose file that contains Nginx, PhpFpm and Varnish.

My Nginx works this way:
User connect to the website => Nginx (443) => Varnish (80) => Nginx (8080) => Phpfpm:9000 or others..

My project work well with "depends_on" config inside docker-compose.
But, with docker swarm, the depends_on is ignored.

There my problems starts..

My Varnish container, needs Nginx to be running, or its crash, due to the hostname defined on top of the configuration file :

````
# varnish config file
backend default {
  .host = "nginx";
  .port = "8080";
  .connect_timeout = 10s;
  .first_byte_timeout = 10s;
  .between_bytes_timeout = 10s;
}
````

And, my nginx, needs varnish to be running or it's crash too...

````
# pass to varnish
location / {
    proxy_pass http://varnish;
}

upstream varnish {
    server varnish:80;
}
````

Soooo, varnish crashs because nginx is not up and nginx crashs because varnish is not up.

Is there any solution for this problem ?

**Comment by BretFisher** [comment]
In distributed computing, you can't assume "one part of the app is up before the other". Each part of your app (each container) needs to have its own way of dealing with unavailable dependent services.  Either retries or just letting docker restart the container until it connects.  I've seen some people add wait-for-it scripts to the ENTRYPOINT but that masks the problem and doesn't deal with "what if I replace the container for one of those services".

I assume you have 4 Swarm Services here? If so, wouldn't they eventually come up, as each starts properly once its backend/upstream is ready?

**Reply by ghostlexly** [reply]
I have also seen some examples with wait-for, but it's can't work on my case, because varnish crash when nginx is down, and nginx crash when varnish is down. :/

**Reply by BretFisher** [reply]
This solution has 4 containers in a docker network right? Why would nginx:8080 crash if Varnish is down?

**Reply by ghostlexly** [reply]
i have 1 nginx, 1 phpfpm and 1 varnish, so 3 containers.
The 8080 and 443 are on the same nginx

**Reply by BretFisher** [reply]
That's what I said previously. Your issue is a circular dependency that is only solved by running two nginx's. See https://github.com/BretFisher/ama/discussions/118

**Reply by BretFisher** [reply]
The setup from #118 is normal. Don't think of it as "2 nginx doing the same thing". You're nginx:8080 is acting as a fastcgi proxy, and your nginx:443 is a tls proxy. Usually those are not the same container.

**Comment by BretFisher** [comment]
Note you should have two independent nginx services there because they are doing different things, and you'll have the circular dependency if you don't split them into different containers.

**Comment by ghostlexly** [comment, marked as answer]
Well, the real issue here is nginx and varnish crashing both forever.
I have finally found a solution by using a variable in proxy_pass, to ignore if the host is down.
But it's not a real solution, it's just a quick fix...
